sillsdev/web-languageforge
# Delta updates to minimize edit conflicts

## Description

When a user edits an entry, the changes are now sent as deltas instead of full entries. This means that instead of replacing the entry entry when the user edits the English definition, that edit will turn into a Mongo update containing just "set definition.en.value = 'new text'". This should greatly reduce the number of times that user edits conflict with each other. When two people work on the same entry at the same time, the only way that their edits can conflict with each other is if they edit the same writing system in the same field.

However, for technical reasons, adding or removing a new Meaning or Example causes a full-entry update rather than a delta update. So if user A adds a new example sentence, this can cause the work that user B had done on the definition to be lost. This should hopefully be a rare scenario, however; we expect that more than 90% of edit conflicts will be solved by this PR.

Fixes #1051 

### Type of Change

Only keep lines below that describe this change, then delete the rest.

- New feature (non-breaking change which adds functionality)

## Screenshots

Please provide screenshots / animations for any change that involves the UI.  Please provide animations to demonstrate user interaction / behavior changes

## How Can This Be Tested?

To test this, two or three people should edit the same entry in the same project. Sit where you can see each others' screens if possible, and watch the edits that the other person is making. Verify the following scenarios:

- [ ] One person edits the lexeme, the other person edits a definition. Both edits go through.
- [ ] One person edits the definition in English, the other person edits the same definition in Spanish. Both edits go through.
- [ ] One person edits the definition, the other edits an example sentence. Both edits go through.
- [ ] One person edits the definition, the other adds a **new** example sentence. This will cause a race condition where sometimes, both edits go through, while at other times, the definition is lost because the example sentence update "squashed" the definition update. (This will usually trigger if the definition update happens first and the new example sentence is saved second, but due to the different speed of the save and refresh timers, it's not always possible to guarantee triggering the race condition). This is a known issue, and would require an unfeasible amount of work to fix.

## Checklist:
